pennylane/wires.py

```diff
@@ -0,0 +1,369 @@
+# Copyright 2018-2020 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+This module contains the :class:`Wires` class, which takes care of wire bookkeeping.
+"""
+from collections import Sequence, Iterable
+import numpy as np
+from numbers import Number
+
+
+class WireError(Exception):
+    """Exception raised by a :class:`~.pennylane.wires.Wire` object when it is unable to process wires.
+    """
+
+
+def _process(wires):
+    """Converts the input to a tuple of numbers or strings."""
+
+    if isinstance(wires, Wires):
+        # if input is already a Wires object, just copy its wire tuple
+        return wires.wire_tuple
+
+    elif isinstance(wires, Iterable) and all(isinstance(w, Wires) for w in wires):
+        # if the elements are themselves Wires objects, merge them to a new one
+        wires = [w for wires_ in wires for w in wires_.tolist()]
+        return tuple(wires)
+
+    elif isinstance(wires, Iterable) and all(isinstance(w, str) or isinstance(w, Number) for w in wires):
+        # if the elements are strings or numbers, turn iterable into tuple
+        return tuple(wires)
+
+    elif isinstance(wires, Number):
+        # if the input is a single number, interpret as a single wire
+        return (wires,)
+
+    else:
+        raise WireError(
+            "Wires must be represented by a number or string; got {} of type {}.".format(
+                wires, type(wires)
+            )
+        )
+
+
+class Wires(Sequence):
+    """
+    A bookkeeping class for wires, which are ordered collections of unique objects. The i'th object
+    addresses the i'th quantum subsystem.
+
+    There is no conceptual difference between registers of multiple wires and single wires,
+    which are just wire registers of length one.
+
+    Indexing and slicing this sequence will return another ``Wires`` object.
+
+    Args:
+         wires (Any): If iterable, interpreted as an ordered collection of unique objects representing wires.
```

Comment:
Just to clarify, this is `wires (Iterable[Number,str],str,Number)`? This will probably be answered below, but judging by the `_process` function above, numbers, strings, and iterables of numbers and strings are the only things allowed?

Reply:
Yes. Made it clearer.